File: akka-docs-new/src/test/java/jdocs/tutorial_5/DeviceGroup.java
/**
 * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package jdocs.tutorial_5;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import scala.concurrent.duration.FiniteDuration;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

public class DeviceGroup extends AbstractActor {
  private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

  final String groupId;

  public DeviceGroup(String groupId) {
    this.groupId = groupId;
  }

  public static Props props(String groupId) {
    return Props.create(jdocs.tutorial_3.DeviceGroup.class, groupId);
  }

  public static final class RequestDeviceList {
    final long requestId;

    public RequestDeviceList(long requestId) {
      this.requestId = requestId;
    }
  }

  public static final class ReplyDeviceList {
    final long requestId;
    final Set<String> ids;

    public ReplyDeviceList(long requestId, Set<String> ids) {
      this.requestId = requestId;
      this.ids = ids;
    }
  }

  public static final class RequestAllTemperatures {
    final long requestId;

    public RequestAllTemperatures(long requestId) {
      this.requestId = requestId;
    }
  }

  public static final class RespondAllTemperatures {
    final long requestId;
    final Map<String, TemperatureReading> temperatures;

    public RespondAllTemperatures(long requestId, Map<String, TemperatureReading> temperatures) {
      this.requestId = requestId;
      this.temperatures = temperatures;
    }
  }

  public static interface TemperatureReading {
  }

  public static final class Temperature implements TemperatureReading {
    public final double value;

    public Temperature(double value) {
      this.value = value;
    }
  }

  public static final class TemperatureNotAvailable implements TemperatureReading {
  }

  public static final class DeviceNotAvailable implements TemperatureReading {
  }

  public static final class DeviceTimedOut implements TemperatureReading {
  }


  Map<String, ActorRef> deviceIdToActor = new HashMap<>();
  Map<ActorRef, String> actorToDeviceId = new HashMap<>();
  long nextCollectionId = 0L;

  @Override
  public void preStart() {
    log.info("DeviceGroup {} started", groupId);
  }

  @Override
  public void postStop() {
    log.info("DeviceGroup {} stopped", groupId);
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
            .match(DeviceManager.RequestTrackDevice.class, trackMsg -> {
              if (this.groupId.equals(trackMsg.groupId)) {
                ActorRef ref = deviceIdToActor.get(trackMsg.deviceId);
                if (ref != null) {
                  ref.forward(trackMsg, getContext());
                } else {
                  log.info("Creating device actor for {}", trackMsg.deviceId);
                  ActorRef deviceActor = getContext().actorOf(Device.props(groupId, trackMsg.deviceId), "device-" + trackMsg.deviceId);
                  getContext().watch(deviceActor);
                  deviceActor.forward(trackMsg, getContext());
                  actorToDeviceId.put(deviceActor, trackMsg.deviceId);
                  deviceIdToActor.put(trackMsg.deviceId, deviceActor);
                }
              } else {
                log.warning(
                        "Ignoring TrackDevice request for {}. This actor is responsible for {}.",
                        groupId, this.groupId
                );
              }
            })
            .match(RequestDeviceList.class, r -> {
              getSender().tell(new ReplyDeviceList(r.requestId, deviceIdToActor.keySet()), getSelf());
            })
            .match(Terminated.class, t -> {
              ActorRef deviceActor = t.getActor();
              String deviceId = actorToDeviceId.get(deviceActor);
              log.info("Device actor for {} has been terminated", deviceId);
              actorToDeviceId.remove(deviceActor);
              deviceIdToActor.remove(deviceId);
            })
            .match(RequestAllTemperatures.class, r -> {
              getContext().actorOf(DeviceGroupQuery.props(
                      actorToDeviceId, r.requestId, getSender(), new FiniteDuration(3, TimeUnit.SECONDS)));
            })
            .build();
  }
}

File: akka-docs/src/test/scala/tutorial_4/DeviceGroup.scala
/**
 * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package tutorial_3

import akka.actor.{ Actor, ActorLogging, ActorRef, Props, Terminated }
import tutorial_3.DeviceGroup._
import tutorial_3.DeviceManager.RequestTrackDevice

import scala.concurrent.duration._

//#device-group-full
//#device-group-register
object DeviceGroup {
  def props(groupId: String): Props = Props(new DeviceGroup(groupId))
  //#device-group-register

  final case class RequestDeviceList(requestId: Long)
  final case class ReplyDeviceList(requestId: Long, ids: Set[String])
  //#device-group-register
}
//#device-group-register
//#device-group-register
//#device-group-remove

class DeviceGroup(groupId: String) extends Actor with ActorLogging {
  var deviceIdToActor = Map.empty[String, ActorRef]
  //#device-group-register
  var actorToDeviceId = Map.empty[ActorRef, String]
  //#device-group-register

  override def preStart(): Unit = log.info("DeviceGroup {} started", groupId)

  override def postStop(): Unit = log.info("DeviceGroup {} stopped", groupId)

  override def receive: Receive = {
    case trackMsg @ RequestTrackDevice(`groupId`, _) =>
      deviceIdToActor.get(trackMsg.deviceId) match {
        case Some(deviceActor) =>
          deviceActor forward trackMsg
        case None =>
          log.info("Creating device actor for {}", trackMsg.deviceId)
          val deviceActor = context.actorOf(Device.props(groupId, trackMsg.deviceId), s"device-${trackMsg.deviceId}")
          //#device-group-register
          context.watch(deviceActor)
          actorToDeviceId += deviceActor -> trackMsg.deviceId
          //#device-group-register
          deviceIdToActor += trackMsg.deviceId -> deviceActor
          deviceActor forward trackMsg
      }

    case RequestTrackDevice(groupId, deviceId) =>
      log.warning(
        "Ignoring TrackDevice request for {}. This actor is responsible for {}.",
        groupId, this.groupId
      )
    //#device-group-register
    //#device-group-remove

    case RequestDeviceList(requestId) =>
      sender() ! ReplyDeviceList(requestId, deviceIdToActor.keySet)
    //#device-group-remove

    case Terminated(deviceActor) =>
      val deviceId = actorToDeviceId(deviceActor)
      log.info("Device actor for {} has been terminated", deviceId)
      actorToDeviceId -= deviceActor
      deviceIdToActor -= deviceId

    //#device-group-register
  }
}
//#device-group-remove
//#device-group-register
//#device-group-full
